[GFC] Move PlacedGridItem construction from ImplicitGrid to GridFormattingContext.

sammygill opened the pull request:
#### efaa47c1f1e0a40e5da3445c01bc82e884f1fc98
<pre>
[GFC] Move PlacedGridItem construction from ImplicitGrid to GridFormattingContext.
<a href="https://bugs.webkit.org/show_bug.cgi?id=299533">https://bugs.webkit.org/show_bug.cgi?id=299533</a>
<a href="https://rdar.apple.com/problem/161337846">rdar://problem/161337846</a>

Reviewed by Brent Fulgham.

Currently, PlacedGridItems are created via placedGridItems() on
ImplicitGrid by constructing them out of the set of UnplacedGridItems
which it has mapped to different areas in the grid.

This has worked well enough for now, but as we prepare to size the
grid items, we need to be able to store additional data on these objects
in order to accomplish this. In particular, we need to store and be able
to use the various sizing properties (width, min-width, max-width, etc.)
for each dimension (inline vs block). However, we do not have enough
information at the ImplicitGrid level in order to map these properties
to the appropriate dimension. Instead, we should have GridFormattingContext
start creating these objects.

Not only will this allow us to correctly create PlacedGridItems with the
data that we need, but it also allows ImplicitGrid to return some simple
structures that describe the ImplicitGrid rather than a more complex
object.

* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::placeGridItems):
(WebCore::Layout::GridLayout::layout):
Previously we were getting the PlacedGridItems as a result of the grid
item placement portion (placeGridItems()) of the layout algorithm. Now
we get the grid areas for each item which were resolved from that
process. We can then ask the GridFormattingContext to create the set of
PlacedGridItems from this data.

* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
(WebCore::Layout::ImplicitGrid::gridAreas const):
Instead of returning PlacedGridItems we return the set of grid areas
contained within the ImplicitGrid. This is just a map of each item that
has been placed in the grid along with the area (set of lines) that it
has been placed in. The caller can then take the UnplacedGridItem and
its grid area and do whatever it needs to which includes constructing a
PlacedGridItem if they need it.

Canonical link: <a href="https://commits.webkit.org/300547@main">https://commits.webkit.org/300547@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ef0a6b4db1d03502273b3f752fc63a89e9d8780

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123050 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42764 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33461 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129706 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/fbe2b95f-c636-4172-81d4-f326784df953) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43488 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51359 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/968a1800-45da-4d48-a444-88a67e102c5f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126001 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/34662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110128 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74164 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e2ae094b-e926-400a-b8f4-ca8f6385a227) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28283 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73218 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/104373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28508 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132436 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50000 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/38060 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102030 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50376 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106348 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/101892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25879 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47266 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/25459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49855 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55616 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/49323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52675 "Built successfully") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/layout/formattingContexts/grid/UnplacedGridItem.h:
```diff
@@ -84,6 +84,7 @@ template<> struct HashTraits<WebCore::Layout::UnplacedGridItem> : SimpleClassHas
     static constexpr bool hasIsEmptyValueFunction = true;
 
     static bool isEmptyValue(const WebCore::Layout::UnplacedGridItem& unplacedGridItem) { return unplacedGridItem.isHashTableEmptyValue(); }
+    static WebCore::Layout::UnplacedGridItem emptyValue () { return WebCore::Layout::UnplacedGridItem { HashTableEmptyValueType::HashTableEmptyValue }; }
```
```suggestion
    static WebCore::Layout::UnplacedGridItem emptyValue() { return WebCore::Layout::UnplacedGridItem { HashTableEmptyValueType::HashTableEmptyValue }; }
```

brentfulgham commented on Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
```diff
@@ -112,5 +113,15 @@ void GridFormattingContext::layout(GridLayoutConstraints layoutConstraints)
     GridLayout { *this }.layout(layoutConstraints, unplacedGridItems);
 }
 
+GridFormattingContext::PlacedGridItems GridFormattingContext::constructPlacedGridItems(GridAreas gridAreas) const
```
Shouldn't you pass the GridAreas HashMap as a const reference (or move if this is not needed after this call?)

sammygill replied:
Yes that should definitely be const ref at least for now! Thanks!

brentfulgham commented on Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.h:
```diff
@@ -51,6 +54,10 @@ class GridFormattingContext : public CanMakeCheckedPtr<GridFormattingContext> {
     void layout(GridLayoutConstraints);
 
+    using PlacedGridItems = Vector<PlacedGridItem>;
+    using GridAreas = HashMap<UnplacedGridItem, GridAreaLines>;
```
Nit: You might want to make a small header file for these types so you don't have to do this declaration in so many places.

sammygill replied:
I think that's a reasonable idea. I am going to do that in a follow up patch so I don't pollute this PR if you don't mind

brentfulgham replied:
Sounds like a good plan!

brentfulgham commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -100,24 +100,21 @@ void ImplicitGrid::insertUnplacedGridItem(const UnplacedGridItem& unplacedGridIt
-    HashSet<UnplacedGridItem> processedUnplacedGridItems;
-    PlacedGridItems placedGridItems;
-
+    GridAreas gridAreas;
```
Nit: We can avoid reallocation/rehashing of GridAreas if you use `reserveInitialCapacity`, maybe setting it to the m_gridMatrix size * the size of the first m_gridMatrix entry? This might imply GridMatrix should be a real type that remembers its member count.

brentfulgham commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -100,24 +100,21 @@ void ImplicitGrid::insertUnplacedGridItem(const UnplacedGridItem& unplacedGridIt
                     continue;
-                processedUnplacedGridItems.add(unplacedGridItem);
-                placedGridItems.append({ unplacedGridItem, { columnIndex, columnIndex + 1, rowIndex, rowIndex + 1 } });
+                gridAreas.add(unplacedGridItem, GridAreaLines { columnIndex, columnIndex + 1, rowIndex, rowIndex + 1 });
```
I think it's more efficient for you to use the `gridAreas.ensure(...)` method. In that method, it attempts to find the entry, and if it is not found uses the computed hash to add the value.

I think this code does two hash operations on `unplacedGridItem` when inserting.

brentfulgham commented on Source/WebCore/layout/formattingContexts/grid/PlacedGridItem.h:
```diff
@@ -33,15 +33,16 @@ namespace Layout {
+    size_t columnEndLine;
+    size_t rowStartLine;
+    size_t rowEndLine;
+};
```
Nit: These should live in a `GridAreaTypes.h` header or something!

brentfulgham reviewed (approved):
The change looks fine, but I think you could improve it in a few ways (as I suggest below). No need for another review from me.

stwrt commented on Source/WebCore/layout/formattingContexts/grid/GridFormattingContext.cpp:
```diff
@@ -112,5 +113,15 @@ void GridFormattingContext::layout(GridLayoutConstraints layoutConstraints)
+    PlacedGridItems placedGridItems;
+    placedGridItems.reserveInitialCapacity(gridAreas.size());
+    for (auto [ unplacedGridItem, gridAreaLines ] : gridAreas)
+        placedGridItems.constructAndAppend(unplacedGridItem, gridAreaLines);
```
Is this copying all the grid items into a new vector?

sammygill replied:
We're going to construct a `PlacedGridItem` out of an `UnplacedGridItem` and `GridAreaLines`. There's might be a more efficient way to do this if needed

brentfulgham replied:
If would be better if we could move the data into the new structure and abandon the old thing. But not needed for this first patch (and perhaps isn't even possible).

brentfulgham commented on Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
```diff
@@ -100,24 +100,25 @@ void ImplicitGrid::insertUnplacedGridItem(const UnplacedGridItem& unplacedGridIt
             for (const auto& unplacedGridItem : gridCell) {
-                if (processedUnplacedGridItems.contains(unplacedGridItem))
+                if (gridAreas.contains(unplacedGridItem))
                     continue;
```
I think you are still double-hashing. The `HashSet::ensure` step will only enter the lambda if the key argument passed is not in the container. So you shouldn't need this early `continue` step, as the call to `gridAreas.ensure(...)` will just immediately bail out after checking if the item is already in the container.

sammygill replied:
Oops you are right! Embarrassing!

sammygill committed:
[GFC] Move PlacedGridItem construction from ImplicitGrid to GridFormattingContext.
https://bugs.webkit.org/show_bug.cgi?id=299533
rdar://problem/161337846

Reviewed by Brent Fulgham.

Currently, PlacedGridItems are created via placedGridItems() on
ImplicitGrid by constructing them out of the set of UnplacedGridItems
which it has mapped to different areas in the grid.

This has worked well enough for now, but as we prepare to size the
grid items, we need to be able to store additional data on these objects
in order to accomplish this. In particular, we need to store and be able
to use the various sizing properties (width, min-width, max-width, etc.)
for each dimension (inline vs block). However, we do not have enough
information at the ImplicitGrid level in order to map these properties
to the appropriate dimension. Instead, we should have GridFormattingContext
start creating these objects.

Not only will this allow us to correctly create PlacedGridItems with the
data that we need, but it also allows ImplicitGrid to return some simple
structures that describe the ImplicitGrid rather than a more complex
object.

* Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
(WebCore::Layout::GridLayout::placeGridItems):
(WebCore::Layout::GridLayout::layout):
Previously we were getting the PlacedGridItems as a result of the grid
item placement portion (placeGridItems()) of the layout algorithm. Now
we get the grid areas for each item which were resolved from that
process. We can then ask the GridFormattingContext to create the set of
PlacedGridItems from this data.

* Source/WebCore/layout/formattingContexts/grid/ImplicitGrid.cpp:
(WebCore::Layout::ImplicitGrid::gridAreas const):
Instead of returning PlacedGridItems we return the set of grid areas
contained within the ImplicitGrid. This is just a map of each item that
has been placed in the grid along with the area (set of lines) that it
has been placed in. The caller can then take the UnplacedGridItem and
its grid area and do whatever it needs to which includes constructing a
PlacedGridItem if they need it.

Canonical link: https://commits.webkit.org/300547@main
